Access Javac API from a standalone javac jar instead of the default JDK runtime

testforstephen opened the pull request:
This PR tries to introduce a standalone Javac jar from the repository https://github.com/JaroslavTulach/nb-javac for Javac API access within our project. The benefit is that it allows the language server to support latest language features without requiring user to run the language server with the latest JDK.

This PR is an experiment PR, not full ready yet. It faces some issues with new approach, and I open the PR first for suggestions and feedback.
- The project can compile fine with the embedded nb-javac, but get the module conflict issue during runtime. It reports module conflicts between jdk.compiler module and unnamed module from the embedded nb-javac jar. When running it in Eclipse, it throws exception below.
 ```
java.lang.ClassCastException: class com.sun.tools.javac.api.JavacTool cannot be cast to class com.sun.tools.javac.api.JavacTool (com.sun.tools.javac.api.JavacTool is in module jdk.compiler of loader 'app'; com.sun.tools.javac.api.JavacTool is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @d014059)
	at org.eclipse.jdt.core.dom.JavacCompilationUnitResolver.parse(JavacCompilationUnitResolver.java:690)
	at org.eclipse.jdt.core.dom.JavacCompilationUnitResolver.toCompilationUnit(JavacCompilationUnitResolver.java:541)
	at org.eclipse.jdt.core.dom.ASTParser.internalCreateASTCached(ASTParser.java:1299)
	at org.eclipse.jdt.core.dom.ASTParser.lambda$1(ASTParser.java:1178)
	at org.eclipse.jdt.internal.core.JavaModelManager.cacheZipFiles(JavaModelManager.java:5692)
	at org.eclipse.jdt.core.dom.ASTParser.internalCreateAST(ASTParser.java:1178)
	at org.eclipse.jdt.core.dom.ASTParser.createAST(ASTParser.java:918)
	at org.eclipse.jdt.internal.core.CompilationUnit.buildStructure(CompilationUnit.java:184)
	at org.eclipse.jdt.internal.core.Openable.generateInfos(Openable.java:246)
	at org.eclipse.jdt.internal.core.JavaElement.openWhenClosed(JavaElement.java:569)
	at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:292)
	at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:278)
	at org.eclipse.jdt.internal.core.CompilationUnit.getSourceRange(CompilationUnit.java:1087)
```
- The packaged nb-javac only includes the classes of jdk.compiler, no jdk.javadoc classes. To bypass the compilation error, I temporarily comment out the jdk.javadoc classes reference from JavadocConverter.java.
- License review. nb-javac is under GPL-2.0 license, require legal review.

Commit:
Fix test0150 - malformed package declaration

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix issue with isGeneric according to spec

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix test0037 - getKey on JavacTypeVariableBinding

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Fix problemId for lambda without return

- This also fixes the quickfix as a result

eg.

```java
Function<Integer, Integer> func = x -> {
    System.out.println(x);
};
```

becomes

```java
Function<Integer, Integer> func = x -> {
    System.out.println(x);
    return x;
};
```

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Use compiler settings to determine problem id for interface modifiers

Different versions of Java allow different modifiers on interface methods,
so ECJ generates slightly different problem ids for each of these cases.
Use the compiler settings to determine which to use.
This affects the logic of the quick fixes,
so it should fix some jdt-ls test cases.

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix testPatternInstanceOfExpression003

Signed-off-by: Rob Stryker <stryker@redhat.com>

akurtakov committed:
Fix ASTConverter_RecordPattern_Test.testCaseDefaultExpressionPattern

mickaelistria committed:
Avoid requesting getKey early: store bindings identity with
hashcode/equals

mickaelistria committed:
Improve JavacBindingResolver.resolveConstantExpression

Commit:
Fix some generic/parameterized in JavacTypeBinding

Commit:
Attempt to fix package bindings

Signed-off-by: Rob Stryker <stryker@redhat.com>

Fix regression

Signed-off-by: Rob Stryker <stryker@redhat.com>

Fix regressions

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fixes testBug497719_0001 - try with resources dom error

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix findDeclaringNode with binding key

Commit:
Partial fix to testBug432175 - keys for types not under main type of compilation unit

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
More problems to ignore

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[Javadoc] error handling

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix equality for generic bindings always hosting generic type

testforstephen committed:
Handle the nested javadoc syntax '@see {@link ...}' better (#705)

Commit:
Fix testRecord012 - synthetic flag

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix testRecordPattern003

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix test0205 relating to source position of method name in annotation type

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix testRecord011 - setting start position of record name

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Different impl for detecting secondary types when getting key

- This implementation doesn't rely on using the javac to dom map,
  which may not be built out yet if one of the referenced types hasn't
  been properly parsed yet

Signed-off-by: David Thompson <davthomp@redhat.com>

testforstephen committed:
Fix NPE when hovering at the keyword 'void' (#712)

mickaelistria committed:
Store methodBindings by hash

instead of using `getKey`

mickaelistria committed:
Add isDeclaration to JavacMethodBinding

So it can be used to return proper declaring class.

akurtakov committed:
Trim o.e.jdtcore.tests.javac dependencies

mickaelistria committed:
Specifically handle equals/hashCode for MethodType

This type can have multiple equivalent instances and is missing
equals/hashCode in the definition. Add our own hash/equals logic.

mickaelistria committed:
Fix position for method identifier

datho7561 committed:
Use absolute path when parsing a file in a project

We were passing the project-relative path of a file to javac when working with Java files
in Java projects.
The reason it was working before was because we were directly sending
the file's contents to the compiler.

This change fixes an error marker I get when opening any module-info.java,
where javac would complain that the file is not on the source path.
This is because javac was told the file was located at `/src/module-info.java`,
but the source path it got was correctly set to (eg.)
`/home/davthomp/Projects/module-project/src`.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix isCastCompatible

Reorder args as doc for types.isCastable says "can t be cast to s?"

mickaelistria committed:
Avoid creating binding for incomplete type objects

mickaelistria committed:
Workaround some cases where Javac symbols are not usable

Eg when trying to get symbol for `UnknownType<TypeParam>`, javac doesn't
recover it well

mickaelistria committed:
Map ambiguous interface problem

datho7561 committed:
Various fixes in order to get lemminx compiling using the builder

- Port over the relative -> absolute path conversion from the JavacCompilationUnitResolver
- Add a logging statement for when the compilation crashes
- Add a null guard to prevent compilation from crashing
- Add an error id conversion I encountered along the way

Signed-off-by: David Thompson <davthomp@redhat.com>

akurtakov committed:
Install o.e.jdt.core.tests.model artifact

Should make tests run with locally built artifacts and not with I-build
ones.

mickaelistria committed:
Fix diagnostic & isRecovered for missing package

mickaelistria committed:
Set range for mock name expressions

when mapping a syntax error.

This should fix some SignatureHelpHandlerTests

mickaelistria committed:
Workaround binding resolution for Arrays.<Number>asList(unknown)

mickaelistria committed:
Fix resolution of method bindings for erroneous generics

datho7561 committed:
Fix `ClassCastException` when working with module projects

- Remove `JavacFileObject`, use file objects from file manager

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Exclude missing modifiers implied by Javac

mickaelistria committed:
Implement JavacVariableBinding.isRecordComponent

mickaelistria committed:
Fix resolveName() in case name doesn't have symbol

mickaelistria committed:
Alphabetically order getDeclaredMethods()

as expected by JDT-LS tests, similar order as with ECJ

mickaelistria committed:
Fix problem range for reduced visibility

mickaelistria committed:
Fix binaryName, better map and reconcile abstract related pb.

robstryker committed:
Fix javadoc test 001 and others (#718)

* Fix javadoc test 001 and others

Signed-off-by: Rob Stryker <stryker@redhat.com>

* Fix javadoc test000

Signed-off-by: Rob Stryker <stryker@redhat.com>

* More work on javadoc tests

Signed-off-by: Rob Stryker <stryker@redhat.com>

* More fixes to javadoc

Signed-off-by: Rob Stryker <stryker@redhat.com>

* Substantial fix for testBug54424, and fixes several unknown others

Signed-off-by: Rob Stryker <stryker@redhat.com>

---------

Signed-off-by: Rob Stryker <stryker@redhat.com>
Co-authored-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Translate some problem IDs for module-related problems

- Fix quickfix for "Create provider method" by adjusting the
  error range of
  "service implementation missing no-args constructor or provider method"
  to match that of ECJ's

to see this quickfix in action, setup the following three files and open
quickfixes on the error in the `module-info.java`.

```java
module myModular {
  provides test.IFoo with test.Foo;
}
```

```java
package test;
public interface IFoo{}
```

```java
package test;
public class Foo implements test.IFoo {
  public Foo(String asdf) {}
}
```

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Improve name binding resolution

Sometimes, the name has no symbol attached, while the "main" parent that
is defined by the name as one.
So we try to resolve parent structure when we couldn't resolve on name.

Commit:
More slight javadoc work, annotating javadoc tests

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix binary name

binary name is only equivalent to signature for primitive types. For
regular named types, the flatname is better

datho7561 committed:
Fixes to switch expressions

- Fix a small bug in the AST
- Fix some problem ids
- Fix some diagnostic ranges

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fixes incubator #740 - ignore some specific malformed declarations

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Work on javadoc inline tags

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Mark fake id as MALFORMED

mickaelistria committed:
Improve JavacTypeBinding.getJavaElement()

Better resolve hosting compilationUnit

mickaelistria committed:
Honor test flag on classpathEntry

mickaelistria committed:
Trigger downstream jdt-ls-javac job upon completion

mickaelistria committed:
Don't wait/propagate downstream job result

Commit:
Fix javadoc bug 68025 for erroneous see tags

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fixes testBug68726 - bad regex for splitting javadoc comments

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix JavacUtils.isTest()

mickaelistria committed:
Do not JavacBindingResolver if not necessary (frees some memory)

mickaelistria committed:
Force resolution of bindings when computeProblems

mickaelistria committed:
Avoid potential NPEs

Commit:
Fix testBug347100

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Cleanup and a small fix

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Revert "Fixes incubator #740 - ignore some specific malformed declarations"

This reverts commit cadf30d278473923f28a28db7b2e56281a42aacb.

mickaelistria committed:
Trim recovered/invalid nodes when recovery is not enabled

mickaelistria committed:
JavacTypeBinding.getJavaElement() returning null for .class with sources

mickaelistria committed:
Prevent NPE for virtual IJavaProject (without IProject)

mickaelistria committed:
Replaced MALFORMED by RECOVERED

Was forgotten in previous commit

Commit:
Fix several tests with malformed package declarations

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Resolve `var` type binding to actual resolved type

mickaelistria committed:
Fix NPE when trimming recovered nodes

mickaelistria committed:
Fix translating empty name to _

mickaelistria committed:
Improve range for variable declaration fragment

Try skipping potential trailing content.

testforstephen committed:
Enable lombok annotation processing when parsing DOM for workingcopy (#759)

mickaelistria committed:
Make JavacBindingResolver more thread-safe

+ Call analyze with FORCE_PROBLEM_DETECTION even without bindings, but
without storing the context
+ prevents for resolving/analyzing more bindings than
necessary

mickaelistria committed:
Fix resolveBindings(LambdaExpression)

Commit:
test annotations

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Improve JavacLambdaBinding and lambda params

* Keep ref to lambda declaration node
* Fix getJavaElement for lambda
* Fix getJavaElement for lambda params

Commit:
Fix test83804 and test83804a - incorrect source ranges / lengths

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Mostly fix test109

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
More Javadoc lint

mickaelistria committed:
Add support for more Javadoc problems

mickaelistria committed:
Fix converting Javadoc for type parameter

mickaelistria committed:
Map javadoc missing return tag

testforstephen committed:
Skip constructor & serialVersionUID & exception parameter of catch block in the unused checker (#774)

* Skip constructor and serialVersionUID field in the unused checker

* Skip the exception parameters in catch block during unused checker

mickaelistria committed:
Improve problem mapping for EnumAbstractMethodMustBeImplemented

mickaelistria committed:
A workaround to resolve constructor binding

When Javac doesn't attach a symbol

mickaelistria committed:
Better recover not visible parameterized types

Dig deeper in the type/symbols to find a usable one

testforstephen committed:
Fix unused private constructor (#777)

mickaelistria committed:
Inspect nested diagnostic in some cases

It may provide finer grain information (eg allows to distinguish permits
vs sealed issues)

mickaelistria committed:
Improve problem conversion to UndefinedName

And trim out some non-compatible args from problem arguments.

Commit:
Fix testBug113108b - three slash comments should be Line Comments

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix testBug87845 - varargs in javadoc method ref parameters

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Ignore testBug103304 - violates javadoc spec

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix converting problems about accessing static members

Commit:
Fix testBug79809 - type parameter javadoc test; previous fix insufficient

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Map error for diamond op on non generic type

eg.
```java
Object asdf = new Object<>();
```

Also fixes the quickfix to remove it.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Better map UndefinedConstructor problem

mickaelistria committed:
Improve JavacBindingResolver.resolveExpressionType()

datho7561 committed:
Fix some more error ids

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Some more problem ids

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Better map missing method diagnostic and related bindings

mickaelistria committed:
Fix potential exception in JavacTypeBinding

mickaelistria committed:
Use some TreeInfo utils instead of custom code

mickaelistria committed:
Bump parent pom version to 4.34

datho7561 committed:
Another round of problem id mappings

This time a lot of them are for errors that ECJ doesn't support

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix type name position

mickaelistria committed:
Add members to ImplicitTypeDeclaration

akurtakov committed:
Map compiler.warn.try.explicit.close.call problem

Commit:
Fix testBug80257 and possibly others: resolve correct type binding in rare case

Signed-off-by: Rob Stryker <stryker@redhat.com>

akurtakov committed:
Code cleanups

testforstephen committed:
Use variable kind to check if it qualifies for unused checker (#801)

mickaelistria committed:
Fix text elements in Snippets

mickaelistria committed:
Fix conversion of Javadoc @value tag

mickaelistria committed:
Fix JavacVariableBinding.getJavaElement() to try fields first

datho7561 committed:
Some more problem IDs

Signed-off-by: David Thompson <davthomp@redhat.com>

testforstephen committed:
Force the DOM parser to continue with errors (#808)

* Force the DOM parser to continue with compilation errors

mickaelistria committed:
Use JDK's snippet markup parser

mickaelistria committed:
Prefer TextElement to Name for @uses

as expected by SemanticTokensHandlerTest.testSemanticTokens_Modules

mickaelistria committed:
Revert "Force the DOM parser to continue with errors (#808)"

The commit has caused several regressions for which a fix is yet to
be found. Let's revert to keep the main branch as clean as possible
and work on TolerantParser in a different PR.

This reverts commit cb3c54ff805b938febe42cabb52e760df1614b29.

mickaelistria committed:
Improve method bindings generic/type/raw

mickaelistria committed:
Try should-stop.ifError Javac options when getting DOM

Skip in case we're running tests to remain compatible with JDT

mickaelistria committed:
Avoid failure in asMethodType()

mickaelistria committed:
Add textElement to DCUses

Commit:
Make small enhancements to jdt.core test suite to allow for proper filtering

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix NPE in JavacMethodBinding

mickaelistria committed:
Add config to use internal Javadoc API in tests

mickaelistria committed:
Avoid NPE in JavacBindingResolver

testforstephen committed:
Configure processor arguments (e.g. -Akey=value) to Javac compiler (#820)

mickaelistria committed:
Make bindings.isEqualTo() respect the spec

and work with bindings coming from different resolvers

mickaelistria committed:
Avoid NPE in ProcessorConfig

When Java project is null or "virtual"

mickaelistria committed:
Fix enum component contructor position

Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/596

mickaelistria committed:
Add export for lombok

mickaelistria committed:
Set -proc:only

datho7561 committed:
Another round of problem id mappings

This one deals mostly with explicit this param.

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Make all model test suites recursively filterable

Signed-off-by: Rob Stryker <stryker@redhat.com>

Remove unused import

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Enable the test filtering on the javac test bundle

Signed-off-by: Rob Stryker <stryker@redhat.com>

Try fixing build issue

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix some UnresolvedMethodsTest because of incorrect range

mickaelistria committed:
Improve support for lombok

* analyze before converting DOM, so the new nodes generated by lombok
are present
* Workaround some oddities in code generated by Lombok
* Workaround extra nodes created by Javac analyze
* Exclude generated constructor (always produced by Javac analysis)
* Fix some bindings

mickaelistria committed:
Also produce specific build of o.e.j.core.compiler.batch

Since we have interesting changes in it.

Commit:
Ignore one test via annotation to verify annotation filtering for tests works

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Build full JDT with Javac p2 repo after build of main branch

mickaelistria committed:
Revert "Improve support for lombok"

This reverts commit 27d1ddbb5831f2bc931f7ebb466afc06873e5d2d.
It causes many regressions on JDT-LS.

mickaelistria committed:
Reapply "Improve support for lombok"

This reverts commit 2aa26316f5bc2690fa6225739c1b195d8cdcece5.
Improve Lombok support by running analysis less greedily

mickaelistria committed:
Improve some bindings, and safer convert

mickaelistria committed:
Fix format in p2.inf

requires 1 VM arg per line

datho7561 committed:
Fix AST conversion error when handling type in array dims

eg. preserve the `m[]` in the AST in the following case,
(despite the fact that it's invalid),
since javac does.

```java
int[] foo = new int[m[]];
```

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix underlying issue in test0009 - new array has incorrect diment sion count

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
De-duplicate test-on-javase23 profile

mickaelistria committed:
Adapt to Java 23 changes

mickaelistria committed:
Fix version in ClassFileConstants

Commit:
Add some annotations to tests

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix slimmed down dom trees via focalPoint - test0486 and others

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
More accurate range

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Support conversion of Markdown Javadoc

mickaelistria committed:
Improve diagnostic location for "virtual" nodes

Javac sometimes add virtual nodes to the tree (usually with endPos ==
-1). Ignore then when computing diagnostic position.

mickaelistria committed:
Avoid NPE with treePath

Fix https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/842

mickaelistria committed:
Fix some source ranges because of Lombok

mickaelistria committed:
Fix 0-length or end-of-line diagnostic position

mickaelistria committed:
Do not crash on non-convertible type

Log instead

mickaelistria committed:
Map unknown compiler error by default

Commit:
Ensure linkplain tag is handled properly

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix reference tags with a module reference included

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Revert "Run more tests with Java 23"

This reverts commit 13618d972355e9c3647995c340a63a09fb561506 as
upstream as applied similar/better change when merging Java 23 support

Commit:
Add helpful comments in code, and add test annotations

Signed-off-by: Rob Stryker <stryker@redhat.com>

Add helpful comments in code, and add test annotations

Signed-off-by: Rob Stryker <stryker@redhat.com>

More annotations

Signed-off-by: Rob Stryker <stryker@redhat.com>

More test annotations

Signed-off-by: Rob Stryker <stryker@redhat.com>

Missing import

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Issue when initializer is null or cannot be converted

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Resolve bindings for annotations on type definitions

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Do not show annotations that had errors resolving unless recover-bindings is set to true

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fine-grain details about converting a Name node when there is an error in parsing or resolution

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Small improvement to accuracy of method binding's getKey method

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Some improvements to Markdown rendering

* Treat DCRawText as raw text (don't both splitting lines)
* Some fixes in TagElement range correction

Commit:
Fix bug in problem reporting where end position was before start position

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Possibly fix https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/617

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Ignoring several snippet tests

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix some javadoc issue mapping

testforstephen committed:
Cache JCCompilationUnit from compilation job for reuse in problem converter (#853)

mickaelistria committed:
Set AccDeprected flag at DOM level

Fix https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/394

datho7561 committed:
Completion to add override for inherited method

eg. try completion at the `|` for the following cases:

```java
public Foo {
    |
}
```

```java
public Foo {
    toStr|
}
```

Closes #859

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
ASTConverter15JLS4Test.test0155 through test0160 - boxing and unboxing

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Add some problem mapping

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
More test annotations

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
index out of bounds exception

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
test0050/X.java - Cleaner handling of variable arity signatures

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Ensure parent packages are added to bindings

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Cleanup

Signed-off-by: Rob Stryker <stryker@redhat.com>

Cleanup

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Improvements to variable and method completion in method bodies

- Filter out private variables and methods from parent classes
- Filter out non-static variables and methods when completing inside a
  static method
- Distinguish between field completion and local variable completion
  (these have different icons in Eclipse)
- Set flags (eg. `public`, `protected`, `final`) for the completion items
  (these have different icons in Eclipse)

Fixes #861

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Make MatchLocator capable of working with DOM

Extra strategy in MatchLocator to use DOM, and tweaks in various
Locators to properly handle DOM. Tests in JavaSearchTests help a lot
(currently ~130/270 passing)

It basically clones and adapt some methods and adapt them to DOM,
A new flag is introduced to decide the strategy.

Done-ish (often with remaining issues)
* FieldLocator
* LocalVariableLocator
* MethodLocator
* SuperTypeReferenceLocator
* TypeDeclarationLocator
* TypeParameterLocator
* TypeReferenceLocator
TODO:
* AndLocator
* ModuleLocator
* PackageDeclarationLocator
* PackageReferenceLocator
* OrLocator
* VariableLocator

datho7561 committed:
Handle incomplete qualified names

eg. completion at the `|`

```java
public class HelloWorld {

    public static void main(String... args) {
        HelloWorld.|
    }
}
```

Handles static methods and members, as well as `this`, `super` and `class`.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Completion on `super.`

Fixes #865

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Map a problem id related to `super()`

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Method argument completion

Fill in method arguments with best guesses of the values,
using the existing mechanism.

- Refactor binding collection logic to filter out non-static members
  when applicable

Closes #868

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve annotation completion

- I broke Gayan's existing support in my previous patches (sorry Gayan), this PR fixes it
- Support completing right after `@`
- Improve annotation parameter completion

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Enable completion tests

- Fix some low hanging completion test errors

Signed-off-by: David Thompson <davthomp@redhat.com>

testforstephen committed:
Set debug info parameters (-g) to javac compiler

datho7561 committed:
Fix some completion relevances

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Completion for keywords that can be inserted where statements can

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Add cancellation checks to DOMCompletionEngine

Signed-off-by: David Thompson <davthomp@redhat.com>

testforstephen committed:
Compile the generated sources into the expected target directory (#879)

datho7561 committed:
Use one default completion strategy

- Remove 'suitable binding' completion
- Fixes to package completionsome
  - Skip suggesting the default package
  - Use the same strategy whenever packages are suggested
  - Improve calculation of the prefix to make the matches more accurate
- Potentially address dupliate type completions

Fixes #870

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Share/deduplicate jar contents

Introduce a subclass of JavacFileManager that is capable of caching and
sharing jar contents.
Also ensure some references are removed from context upon usage to
encourage garbage collection.

Note that JDK content is still duplicated as it's using another
filemanager that seems harder to configure, using direct call to
`PlatformUtils.lookupPlatformDescription().getFileManager()`.
Also note that the cached entries only get disposed when all
filemanagers are closed/GCed.

datho7561 committed:
Begin reporting on completion progress monitor

See https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/3167#issuecomment-2407450906

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Mark the completion task as done using the progress monitor

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix test0080 - generics vs raw vs parameterized

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix test0164 - getDeclaringMethod() should return the declaration binding

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Complete method return type, modifiers and thrown exceptions

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve ClassInstanceCreation completion

- recover `new ` in AST converter
- some fixes in bindings to get some test cases working
- scan backwards for first non-whitespace character when performing AST
  search
- use type hierachy to build a list of potential constructors
  - This method is expensive but accurate
- suggest creating anonymous classes when completing constructors for interfaces
- always provide all constructors of current type to be bug compatible
  with existing CompletionEngine
- use diamond operators

Limitations:
- relevance numbers are not quite right yet
- generic types don't quite work as expected

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Prevent error when renaming variable

Use a more advanced strategy when picking out which working copy to remove,
since sometimes the provided name of the file is only the last segment
of the path.

Fixes #884

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve field completion (`name.|`)

- Use bindings to build out a list of possible suggestions
- Handle access modifiers properly (`private`, `static`, `abstract`, ...)
  - Exclude members in grandparent types whose access was narrowed in
    parent types
- Fix completion preceeding another statement eg.

```java
{
  name.|
  Object myObject = null;
}
```
- Fix `super.to|`
- Prevent `super.|` and `super.to|` from suggesting methods that are
  abstract in the parent type

Fixes #891

Signed-off-by: David Thompson <davthomp@redhat.com>

snjeza committed:
JavacUtils.classpathEntriesToFiles doesn't configure dependent projects

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

datho7561 committed:
Fix some NPEs in completion

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Some fixes to keyword completion

- Perform prefix validation for keywords
- Suggest `super` when a statement is expected

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Avoid NumberFormatException when trying to get latest JRE.

akurtakov committed:
Really ignore compiler.warn.dangling.doc.comment

At some point the problem id that get ignored changed from 0 to -1 but
this warning code was missed in the conversion.
This saves a warning in every file with license header.

datho7561 committed:
Improve completion in extends/implements

- Change dom conversion logic to recover incomplete type names properly
- Restrict the suggested classes based on if they are a class or
  interface
- Do not suggest final classes
- Do not suggest the current class

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Prevent exception with multiline Javadoc method ref

Trim off the leading `*` when working on a method reference that spans
multiple Javadoc lines.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fixes to relevance and specific completion content

Makes changes to what fields in completion results we set
and what we set them to,
with the goal of reducing test failures

Signed-off-by: David Thompson <davthomp@redhat.com>

snjeza committed:
Improve JavacBindingResolver

akurtakov committed:
Map "compiler.warn.override.unchecked.ret"

IProblem.UnsafeReturnTypeOverride seems the correct one.

akurtakov committed:
Ignore 2 warnings without positions

These are compiler.note.removal.filename,
compiler.note.deprecated.plural.additional and the fact that they don't
have position info in them makes them not suitable to create Problems
for them.

datho7561 committed:
Fix some errors that occur when importing o.e.j.c incubator

- Do not reuse the context when compiling to multiple output directories
- Implement Alex's idea of a separate file of javac-specific error codes

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Remove false positive 'unused import' for import used in Javadoc

eg.

```java
import java.util.List;

/**
 * Like a {@link List} if it wasn't a list.
 */
public class NotAList {
}
```

The `java.util.List` import is needed to distiguish the Javadoc mention
of `List` from other potential `List` classes.

Signed-off-by: David Thompson <davthomp@redhat.com>

akurtakov committed:
Fix StringIndexOutOfBoundsException in Javadoc conversion

Make sure that calling substring is done with acceptable values, values
bigger than string length or negative ones are plain wrong.
```
!ENTRY org.eclipse.jdt.core 4 0 2024-10-31 11:45:52.859
!MESSAGE Failed to convert Javadoc
!STACK 0
java.lang.StringIndexOutOfBoundsException: Index 2912 out of bounds for
length 1736
```

datho7561 committed:
Field access and extends/implements completion improvements

- set prefix for `this.t|` completions correctly
- improve extends and implements completion with keywords
eg. avoid generic completion results and instead recommend keywords here:

```java
public class MyClass | {
	// ...
}
```

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
More fixes for field access

- Fix completion issue with `myVariable.|\nMyObj asdf = null;`
- Fix for `myTarget.|` completion

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Better support for ASTParser.setFocalPoint()

Reduce the parsed AST before resolving to get a performance boost.

akurtakov committed:
Fix wrong unused import for throws javadoc

mickaelistria commented:
Does the launch configuration or the java command still set the `--add-exports ...` flag? If so, please make sure they're removed from most places. The META-INF/p2.inf file could probably be removed too.

testforstephen commented:
I removed the `--add-exports ...` flags from the launch configuration. However, Eclipse automatically adds `--add-modules=ALL_SYSTEM` to the launch command, which includes the `java.compiler` module in the runtime by default. This causes the call to `ToolProvider.getSystemJavaCompiler()` to use the default `java.compiler` api and return the default JDK compiler instance and its associated jdk.compiler module. Since nb-javac already contains all the classes from both `java.compiler` and `jdk.compiler`, we need to find a way to ensure that the nb-javac JAR is prioritized over the JDK module.

https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/blob/d8b4f126e4ff2db277ca7715cd3083cd1eef6fd2/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacCompilationUnitResolver.java#L565

akurtakov committed:
Do not generate false positive unused parameter warning

The following cases are fixed:
* Interface methods
* Abstract methods
* Methods marked with Override annotation

datho7561 committed:
Fix some regressions caused by field access fix

- `myMethod().|`
- completion after method declaration in `@interface`

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Share the FileManager used when -release is set

datho7561 committed:
Manage prefix for extends and implements completion

Recover the prefix in this case:

```java
public class MyClass ext| {
}
```

So that we know to only suggest `extends` and not `implements`.

Also, do not suggest `implements` for interfaces.

Signed-off-by: David Thompson <davthomp@redhat.com>

rgrunber commented:
Found this on ClearlyDefined with a score that makes it usable, ~~https://clearlydefined.io/definitions/sourcearchive/mavencentral/com.dukescript.nbjavac/nb-javac/jdk-19+33~~ (not useable, License score is what counts, and it's 25, must be at least 60) . I have filed https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/17140 for the particular version here.

**Update:** It is Approved :heavy_check_mark:

datho7561 committed:
Improve completion for annotation members

- fix duplicates in annotation members
- fix range for annotation members
- fix type of annotation member completions
- opening completion in a normal annotation or single value annotation
  before `(`, eg `@MyAnnota|tion(value = 1)`,
  now tries to complete the annotation name rather than the annotation
  members

Fixes #930

Signed-off-by: David Thompson <davthomp@redhat.com>

testforstephen commented:
In recent commits, I addressed two issues:
- The first issue is `java.lang.ClassCastException: class com.sun.tools.javac.api.JavacTool cannot be cast to class com.sun.tools.javac.api.JavacTool (com.sun.tools.javac.api.JavacTool is in module jdk.compiler of loader 'app'; com.sun.tools.javac.api.JavacTool is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @d014059)`. When launching Eclipse plugins with JDK 9+, most `java.*` modules (including `java.compiler`) are loaded as root modules by default, which is not configurable. To workaround the module conflicts, I use `com.sun.tools.javac.api.JavacTool.create()` to obtain a compiler instance, rather than `javax.tools.ToolProvider.getSystemJavaCompiler()`.

- The second issue is that [nbjavac.VMWrapper](https://github.com/JaroslavTulach/nb-javac/blob/master/make/langtools/netbeans/nb-javac/src/nbjavac/VMWrapper.java#L89) fails to resolve ct.sym from the OSGi classloader. The nb-javac library includes wrappers to load JDK resources, relying on the classloader mechanism for this purpose. However, its implementation assumes that NetBeans loads the nbjavac JARs as standard jar dependencies on the classpath. In the Eclipse environment, where the javac JAR is embedded within a bundle, the classloader behaves differently, causing the nbjavac implementation to fail in an OSGi bundle context. To address this, I implemented a workaround to help VMWrapper load ct.sym from the OSGi bundle resources.

![image](https://github.com/user-attachments/assets/b5f5d626-db50-4f38-800e-41292d138129)

In the latest commit, the DOM resolver already works with the embedded Javac. But for the build support, it still has some problems because the apt plugin is not loaded well. Here are some potential issues to look further.

- The one unresolved issue is that `org.eclipse.jdt.apt.pluggable.core` plugin cannot be compiled well because the module namespace conflicts between the `java.compiler` and unnamed module for the embedded Javac jar. This occurs because the javac plugin is implemented as a fragment host of the `org.eclipse.jdt.core` plugin, causing any plugin that depends on jdt.core to automatically include the javac plugin in its dependencies. As a result, the transitive unnamed nbjavac module in Javac plugin conflicts with the `java.compiler` dependency of the apt plugin. To resolve this, we may need to refactor the alternative compiler API support in `AbstractImageBuilder` into an extension point. This can make the javac plugin into a standalone plugin, and  isolate it from conflicts with other plugins.

![image](https://github.com/user-attachments/assets/6976a791-7ef8-4e63-9897-548d883b5d7b)

- Another potential issue relates to Lombok compatibility. Similar to ECJ, Lombok support can encounter compatibility issues when used with Javac. Just as a new ECJ version might disrupt Lombok functionality, a new Javac release can also break compatibility with Lombok. In jdt.ls, we typically embed the latest lombok.jar and utilize it alongside ECJ. However, with Javac, it defaults to using a project's own Lombok dependency, loading it as a standard annotation processor in Javac.  I recently observed that lombok 1.18.34 is not compatible with javac 23 because of new Javadoc syntax. As a result, applications still using Lombok 1.18.34 will encounter resolution errors with Javac 23. In future, we might consider using the embedded Lombok version in jdt.ls for the Javac parsing as well.

datho7561 committed:
Improve completion for binary infix operator

In order to accomplish this,
I have changed the logic for adjusting the offset used for the node search.

This causes some regressions:
- A few cases where the test expects the AST not to be recovered but javac does recover it (eg. org.eclipse.jdt.core.tests.model.CompletionTests_1_5.test0197)
- @| in Javadoc is not handled. Before it got no completion due to not finding the correct node, but now it gets default completion, which causes some regressions. I'll fix these when dealing with @| properly
- The test cases sometimes expect the current class to be suggested in new | completion even when the type doesn't match, but sometimes they don't. I'll need to investigate further

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
No completion in strings, text blocks, comments, or Javadoc

- Javadoc will need to be revisited, but for now I turned off the
  default completion

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
[WIP] double colon completion

work-in-progress:
- prefix filtering
- see if I can get the binding key bugs resolved

Fixes #938

Signed-off-by: David Thompson <davthomp@redhat.com>

testforstephen commented:
> The one unresolved issue is that org.eclipse.jdt.apt.pluggable.core plugin cannot be compiled well because the module namespace conflicts between the java.compiler and unnamed module for the embedded Javac jar. This occurs because the javac plugin is implemented as a fragment host of the org.eclipse.jdt.core plugin, causing any plugin that depends on jdt.core to automatically include the javac plugin in its dependencies. As a result, the transitive unnamed nbjavac module in Javac plugin conflicts with the java.compiler dependency of the apt plugin. To resolve this, we may need to refactor the alternative compiler API support in AbstractImageBuilder into an extension point. This can make the javac plugin into a standalone plugin, and isolate it from conflicts with other plugins.

I attempted to convert the `org.eclipse.jdt.core.javac` bundle from a fragment to a plugin and used an extension point to register an alternative Javac compiler with AbstractImageBuilder. The image builder successfully locates and utilizes JavacCompiler. However, JavacCompilationUnitResolver fails because it depends on several package-private classes (e.g., `org.eclipse.jdt.core.dom.BindingResolver`) from the `org.eclipse.jdt.core` plugin.
```
java.lang.IllegalAccessError: class org.eclipse.jdt.core.dom.JavacBindingResolver cannot access its superclass 
org.eclipse.jdt.core.dom.BindingResolver (org.eclipse.jdt.core.dom.JavacBindingResolver is in unnamed module of loader 
org.eclipse.osgi.internal.loader.EquinoxClassLoader @5f618123; org.eclipse.jdt.core.dom.BindingResolver is in unnamed module 
of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @90ba3b1)
```


I guess this dependency issue is why the javac plugin is set up as a fragment of jdt.core, allowing it to share the same classloader.

mickaelistria commented on org.eclipse.jdt.core.javac/META-INF/MANIFEST.MF:
```diff
@@ -5,7 +5,10 @@ Bundle-SymbolicName: org.eclipse.jdt.core.javac;singleton:=true
 Fragment-Host: org.eclipse.jdt.core
 Automatic-Module-Name: org.eclipse.jdt.core.javac
-Require-Capability: osgi.ee; filter:="(&(osgi.ee=JavaSE)(version=23))"
+Require-Capability: osgi.ee; filter:="(&(osgi.ee=JavaSE)(version=17))"
```
See https://ci.eclipse.org/ls/job/jdt-core-incubator/job/PR-924/3/console , this breaks current compilation since some Java 21 features are already adopted in the code. Would 21 be fine?

testforstephen replied:
yes, it breaks Java 21 usage (e.g. some utility of String and pattern matching). We need to rewrite them with Java 17 compatible way. The purpose here is to make the minimum JDK requirement same as other JDT plugins.

mickaelistria replied:
IMO, at this stage, we should minimize the changes in the code and stick to Java 21 to facilitate progress. When we've adopted nb-javac, then we'll be more easily able to move code back to Java 17. Moreover, m2e already requires Java 21, so JDT-LS already requires Java 21, so trying to go below that won't produce any benefit for JDT-LS nor vscode-java.

testforstephen replied:
I see redhat.java extension still requires a minimum Java 17 because it stays at m2e@2.6.0. Good to know we're going to upgrade to Java 21.

mickaelistria committed:
Update doc

mickaelistria committed:
Restore completion strategy switch

testforstephen commented:
By default, the JRE takes precedence if the same package exists both in your own jar and JRE. I took a look at how NetBeans makes the local nbjavac jar takes precedence over the JRE. It turns out that NetBeans platform offers the [Ability to hide JRE packages from the module classpath](https://bits.netbeans.org/dev/javadoc/org-openide-modules/apichanges.html#:~:text=is%20not%20public.-,Ability%20to%20hide%20JRE%20packages%20from%20the%20module%20classpath,-Feb%2023%20%2708), that supports using an attribute [OpenIDE-Module-Hide-Classpath-Packages:](https://github.com/apache/netbeans/blob/master/java/libs.nbjavacapi/manifest.mf) in manifest file to hide some JRE packages. This allows a NetBeans module to overlap some JRE packages with own versions.

@mickaelistria Do you know if Eclipse platform provides a similar capability?

datho7561 committed:
Handle resolving bindings for generic constructors

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Completion for Javadoc tags

- Both block and inline tags
- Take into account which ones are applicable to the declaration being
  Javadocumented

Fixes #948

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Ignoring Javadoc of depedencies

When resolving, we do only need the symbols/bindings of depedencies.
Minimize the amount of option to skip eg parsing Javadoc for transitive
files that are only relevant as part of binding resolution.

mickaelistria committed:
Prevent from closing the Platform FileManager

as it's reused on the whole lifecyle of JDT.

mickaelistria committed:
Reduce processing using -implicit:none

* Also do not share the Platform filemanager with build (a bit more
isolation)
* a few other optimization to save CPU on irrelevant operations

mickaelistria committed:
Add output folders to classpath

Because resolving from .class deps is much faster than resolving from
.java deps

mickaelistria committed:
Skip linting without binding nor FORCE_PROBLEM_DETECTION

It's most likely useless to get linting when none of those are
requested.

testforstephen committed:
Access Javac API from a standalone Javac jar insteadof the JDK

testforstephen commented:
Following the suggestions in https://github.com/eclipse-equinox/equinox/discussions/697, I have removed the `Require-Bundle` for the dependency `org.eclipse.core.runtime` from `org.eclipse.jdt.core` and replace it with `Import-Package`. This change allows the locally embedded javac JAR packages to take precedence over the JRE.

However, one remaining issue with the embedded javac is that it does not include the `jdk.javadoc.*` packages. To render snippets for Java 23 Markdown Javadoc, we rely on classes from these packages, such as `jdk.javadoc.internal.doclets.formats.html.taglets.snippet`. Since these classes were introduced in JDK 23 and are missing in the embedded javac JAR, I have commented out this functionality in the PR to ensure the embedded javac passes compilation.

If we choose to support the embedded javac library, we will need to explore alternative approaches to implement snippet support.

akurtakov commented:
IMO it should be reported to netbeans and hear from them whether they are willing to include these packages before making a decision.

testforstephen commented:
> IMO it should be reported to netbeans and hear from them whether they are willing to include these packages before making a decision.

`jdk.javadoc` is typically used by the `javadoc` tool. Could anyone provide more context to explain why we need to package it to a javac library?

mickaelistria commented:
> jdk.javadoc is typically used by the javadoc tool. Could anyone provide more context to explain why we need to package it to a javac library?

Eclipse JDT provides semantic search/reference/bindings/completion... inside Javadoc. In JDT DOM, Javadoc are pretty standard elements and the javadoc needs to be parsed and its symbols resolved too in order to enable all the expected features. The jdk.javadoc package provides some APIs to retrieve the necessary info from JDK and turn it into JDT model.

testforstephen commented:
> Eclipse JDT provides semantic search/reference/bindings/completion... inside Javadoc. In JDT DOM, Javadoc are pretty standard elements and the javadoc needs to be parsed and its symbols resolved too in order to enable all the expected features. The jdk.javadoc package provides some APIs to retrieve the necessary info from JDK and turn it into JDT model.

I think you mentioned here is the javadoc comments in source code. The [com.sun.source.doctree](https://docs.oracle.com/en/java/javase/23/docs/api/jdk.compiler/com/sun/source/doctree/package-summary.html) package in javac (part of the `jdk.compiler` module) already provides interfaces for parsing these comments. And the existing nb-javac library can handle it as well.

On the other hand, [jdk.javadoc](https://docs.oracle.com/en/java/javase/21/docs/api/jdk.javadoc/module-summary.html) module is primarily used by `javadoc` CLI to generate HTML documentation from the javadoc comments in the source code. This is a separate tool from javac, which is why I doubt to ask the NetBeans to package it to the javac library.

mickaelistria commented:
I had missed on message, sorry. So it looks like the Javadoc snippet feature requires the jdk.compiler module. In practice, I don't think there is a way to provide that without using JDK stuff as AFAIK there is really not any other existing implementation.

testforstephen commented:
it's this PR https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/812 that uses some classes from jdk.javadoc. I see you use it to parse markdown/html. I believe there is alternative open-source markdown library we can use for this.

mickaelistria commented:
I'm wondering whether it wouldn't be simpler to just ship 1 fragment for each compatible version of Java that to use nb-javac. The only drawback would be that when a newer version of Java appears, we'd have to create a new fragment for it and ensure some compatibility (this can be configured in a build to ensure we use the right Java version); but even that drawback can IMO lead to higher quality on the long run.

testforstephen commented:
> I'm wondering whether it wouldn't be simpler to just ship 1 fragment for each compatible version of Java that to use nb-javac. The only drawback would be that when a newer version of Java appears, we'd have to create a new fragment for it and ensure some compatibility (this can be configured in a build to ensure we use the right Java version); but even that drawback can IMO lead to higher quality on the long run.

what benefits can we get from packaging multiple fragments for different java versions? could you explain it more?

mickaelistria commented:
> what benefits can we get from packaging multiple fragments for different java versions? could you explain it more?

We could get Javac stuff in JDT-LS running on any current or older version we decide to target.

testforstephen commented:
> We could get Javac stuff in JDT-LS running on any current or older version we decide to target.

this can be a way to solve the compatibility issue of the javac plugins such as lombok.

testforstephen commented:
> However, one remaining issue with the embedded javac is that it does not include the jdk.javadoc.* packages. To render snippets for Java 23 Markdown Javadoc, we rely on classes from these packages, such as jdk.javadoc.internal.doclets.formats.html.taglets.snippet. Since these classes were introduced in JDK 23 and are missing in the embedded javac JAR, I have commented out this functionality in the PR to ensure the embedded javac passes compilation.

For snippet javadoc issue in the PR https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/812, this is just a formatting thing. I'm not too worried about its support when we switch to nb-javac. We can implement a simple algorithm to handle it ourselves or reuse the solution of ECJ such as https://github.com/eclipse-jdt/eclipse.jdt.core/pull/60.

testforstephen committed:
Decouple javac libraries into a separate bundle for better isolation

mickaelistria commented:
For the license, here is the hint from EDP handbook: https://www.eclipse.org/projects/handbook/#ip-sbom . We should probably add an `about.html` mentioning the 3rd party GPLv2+ClassPathExtension license too.

rgrunber commented:
See https://github.com/eclipse-orbit/orbit/blob/main/glassfish/hk2/org.glassfish.hk2.api_2.6.1/src/main/resources/about.html , https://github.com/eclipse-orbit/orbit/blob/main/javax/javax.ws.rs_2.1.6/src/main/resources/about.html for some examples of this.

testforstephen commented:
Split javac jar reference into a separate PR https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/1082, and add `about.html` for license explanation.
